### PR TITLE
[frontent-api] GQL resolvers refactoring

### PR DESCRIPTION
**Pull request by RostislavKreisinger**
| Q             | A
| ------------- | ---
|Description, reason for the PR| For better developer experience we want GQL queries and mutations aliases to be named same as PHP methods in Query and Mutation classes. Using abstract class for query and mutation (see our documentation) it is not necessary to create aliases manually.
|New feature| Yes 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| Yes 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 1]
```diff
@@ -160,36 +160,53 @@ Query:
...
-
-There are several ways how to define resolvers in definition YAML files.
-We use `Overblog\GraphQLBundle\Definition\Resolver\AliasedInterface` to keep definitions simple and easy to read.
-
-`AliasedInterface` describe one method `getAliases` that should return array of `"method name" => "name used in definition"` pairs.
+### Queries
+Queries are normal Symfony services.
+They only have to extend `Shopsys\FrontendApiBundle\Model\Resolver\AbstractQuery` to be recognized as an available query for the GraphQL.
+This abstract class provides convenience for registering query functions by introducing a function naming convention for use in Query.types.yaml definitions. 
+The suffix `Query` in the name of the function is required for calling function as the alias in query decorator (`resolve: "@=query('categoriesQuery')"`).
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md:167]:
```suggestion
This abstract class provides automatic registering of methods ending with `Query` e.g. `categoriesQuery` as aliases, so you can then use method name as `query` (`resolve: "@=query('categoriesQuery')"`) in Query.types.yaml definitions. 
```

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 2]
```diff
@@ -160,36 +160,53 @@ Query:
...
-There are several ways how to define resolvers in definition YAML files.
-We use `Overblog\GraphQLBundle\Definition\Resolver\AliasedInterface` to keep definitions simple and easy to read.
-
-`AliasedInterface` describe one method `getAliases` that should return array of `"method name" => "name used in definition"` pairs.
+### Queries
+Queries are normal Symfony services.
+They only have to extend `Shopsys\FrontendApiBundle\Model\Resolver\AbstractQuery` to be recognized as an available query for the GraphQL.
+This abstract class provides convenience for registering query functions by introducing a function naming convention for use in Query.types.yaml definitions. 
+The suffix `Query` in the name of the function is required for calling function as the alias in query decorator (`resolve: "@=query('categoriesQuery')"`). 
+For example, if you want to get all categories in the GQL query categories, the name of the function will be `categoriesQuery`.
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md:168]:
```suggestion
For example, if you want to get all categories in the GQL query categories, the name of the function should be `categoriesQuery`.
```

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 3]
```diff
@@ -160,36 +160,53 @@ Query:
...
 
-/**
- * @return string[]
- */
-public static function getAliases(): array
+### Mutations
+Mutations are normal Symfony services.
+They only have to extend `Shopsys\FrontendApiBundle\Model\Mutation\AbstractMutation` to be recognized as an available mutation for the GraphQL.
+This abstract class provides convenience for registering mutation functions by introducing a function naming convention for use in Mutation.types.yaml definitions.
+The suffix `Mutation` in the name of the function is required for calling function as the alias in query decorator (`resolve: "@=mutation('createOrderMutation', args, validator)"`).
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md:189]:
```suggestion
This abstract class provides automatic registering of methods ending with `Mutation` e.g. `createOrderMutation ` as aliases, so you can then use method name as `mutation` (`resolve: "@=mutation('createOrderMutation', args, validator)"`) in Mutation.types.yaml definitions. 
```

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 4]
```diff
@@ -160,36 +160,53 @@ Query:
...
+    {
+        // implementation
+
+        return $order;
+    }
 }
 ```
 
-You can override resolvers like any other Symfony service.
+You can override mutations like any other Symfony service.
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md:209]:
```suggestion
You can override both queries and mutations like any other Symfony service.
```

**Review by TomasLudvik — CHANGES_REQUESTED** [review]

File: docs/frontend-api/introduction-to-frontend-api.md (hunk 4 above)
Comment (RostislavKreisinger) [reply docs/frontend-api/introduction-to-frontend-api.md:209]: ok

File: docs/frontend-api/introduction-to-frontend-api.md (hunk 3 above)
Comment (RostislavKreisinger) [reply docs/frontend-api/introduction-to-frontend-api.md:189]: i thing my description is better and more descriptive.

File: docs/frontend-api/introduction-to-frontend-api.md (hunk 1 above)
Comment (RostislavKreisinger) [reply docs/frontend-api/introduction-to-frontend-api.md:167]: i thing my description is better and more descriptive.

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 5]
```diff
@@ -149,7 +149,7 @@ QueryDecorator:
         fields:
             categories:
                 type: '[Category!]!'                    # Array of the categories will be returned.
-                resolve: "@=query('categories')"   # Define the resolver responsible for returning the data. See the resolvers section below.
+                resolve: "@=query('categoriesQuery')"   # Define the query responsible for returning the data. See the 'Queries' section below, ex. CategoriesQuery::categoriesQuery.
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply docs/frontend-api/introduction-to-frontend-api.md:152]: ok

File: packages/frontend-api/src/Resources/config/graphql-types/AdvertImageDecorator.types.yaml [hunk 6]
```diff
@@ -9,9 +9,9 @@ AdvertImageDecorator:
                 type: "String"
                 description: "Advert link"
             image:
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply packages/frontend-api/src/Resources/config/graphql-types/AdvertImageDecorator.types.yaml:11]: ok

File: packages/frontend-api/src/Model/Resolver/Image/ImagesQuery.php [hunk 7]
```diff
@@ -87,9 +60,11 @@ public function resolveByProduct($data, ?string $type, ?string $size): array
      * @param string|null $size
      * @return array
      */
-    public function resolveByCategory(Category $category, ?string $type, ?string $size): array
+    public function imagesByCategoryQuery(Category $category, ?string $type, ?string $size): array
     {
-        return $this->resolveByEntityId($category->getId(), static::IMAGE_ENTITY_CATEGORY, $type, $size);
+        $entityName = $this->imageConfig->getEntityName($category);
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply packages/frontend-api/src/Model/Resolver/Image/ImagesQuery.php:65]: as a programmer I don't want to do double work, the definitions are in images.yaml and I want to consider that as the only source of truth. As far as speed and expandability are concerned, I don't see a problem.

File: packages/frontend-api/src/Model/Resolver/AbstractQuery.php [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+namespace Shopsys\FrontendApiBundle\Model\Resolver;
+
+use Overblog\GraphQLBundle\Definition\Resolver\AliasedInterface;
+use Overblog\GraphQLBundle\Definition\Resolver\QueryInterface;
+use ReflectionClass;
+use ReflectionMethod;
+
+abstract class AbstractQuery implements AliasedInterface, QueryInterface
+{
+    public const QUERY_SUFFIX = 'Query';
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply packages/frontend-api/src/Model/Resolver/AbstractQuery.php:14]: ok

File: docs/frontend-api/introduction-to-frontend-api.md (hunk 2 above)
Comment (RostislavKreisinger) [reply docs/frontend-api/introduction-to-frontend-api.md:168]: I used 'have to' instead

File: packages/frontend-api/src/Resources/config/services.yaml [hunk 9]
```diff
@@ -5,7 +5,7 @@ services:
         public: false
 
     Shopsys\FrontendApiBundle\:
-        resource: '../../**/*{Facade,Factory,Mapper,Mutation,Repository,Resolver,Validator}.php'
+        resource: '../../**/*{Facade,Factory,Mapper,Mutation,Query,Repository,Validator}.php'
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply packages/frontend-api/src/Resources/config/services.yaml:8]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 10]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        ```diff
+        - public function resolve(): array
+        + public function transportsQuery(): array
+        ```
+        - in `Shopsys\FrontendApiBundle\Model\Resolver\Transport\TransportQuery` class is renamed function:
+        ```diff
+        - public function resolver(string $uuid): Transport
+        + public function transportByTransportUuidQuery(string $uuid): Transport
+        ```
+    - in all GQL queries and attribute resolve definitions are replaced all old aliases for renamed Query functions, e.g.:
```
Comment (TomasLudvik) [inline upgrade/UPGRADE-v11.0.0-dev.md:1440]:
```suggestion
    - in all GQL queries resolve definitions are all old aliases replaced by renamed Query functions, e.g.:
```

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 11]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        ```diff
+        - resolve: '@=query("Shopsys\\FrontendApiBundle\\Model\\Resolver\\Image\\ImagesResolver::resolveByAdvert", value, args["type"], args["size"])'
+        + resolve: '@=query("imagesByAdvertQuery", value, args["type"], args["size"])'
+        ```
+    - item in images array cannot be null and array as it is not nullable too. The correction is concerned `AdvertImageDecorator`, `BrandDecorator` and `CategoryDecorator`.
```
Comment (TomasLudvik) [inline upgrade/UPGRADE-v11.0.0-dev.md:1445]:
```suggestion
    - type definition of `images` in `AdvertImageDecorator`, `BrandDecorator` and `CategoryDecorator` has changed:
```

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 12]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        ```diff
+        - type: "[Image]"
+        + type: "[Image!]!"
+        ```
+    - in GQL definition of `AdvertImageDecorator` is fixed attribute for images array.
```
Comment (TomasLudvik) [inline upgrade/UPGRADE-v11.0.0-dev.md:1450]:
```suggestion
    - in GQL definition of `AdvertImageDecorator` is attribute `image` renamed to `images` and its type definition has changed:
```

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 13]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        ```diff
+        - image:
+        -     type: "[Image]"
+        + images:
+        +     type: "[Image!]!"
+        ```
+    - All functions in Mutation classes used as alias are renamed by convention with suffix `Mutation`
```
Comment (TomasLudvik) [inline upgrade/UPGRADE-v11.0.0-dev.md:1457]:
```suggestion
    - All mutation methods in `*Mutation` classes should end with suffix `Mutation` in order to be automatically aliased for use in GQL definitions
```

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 14]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        - in `Shopsys\FrontendApiBundle\Model\Mutation\Order\CreateOrderMutation` class is renamed function:
+        ```diff
+        - public function createOrder(Argument $argument, InputValidator $validator): Order
+        + public function createOrderMutation(Argument $argument, InputValidator $validator): Order
+        ```
+  - in GQL mutation definitions are replaced all old aliases for renamed Mutation functions, e.g.:
+        ```diff
+        - resolve: "@=mutation('create_order', args, validator)"
+        + resolve: "@=mutation('createOrderMutation', args, validator)"
+        ```
```
Comment (TomasLudvik) [inline upgrade/UPGRADE-v11.0.0-dev.md:1500]: As I see this, there is no simple rule to be applied. You have changed names in multiple ways thus I think you should create list of changes

File: packages/frontend-api/src/Resources/config/services.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/frontend-api/src/Resources/config/services.yaml:8]: You have mentioned changes you have done to the names of current classes but you did not instruct developers to do the same for theirs current classes as they will stop autowiring.

**Review by TomasLudvik — CHANGES_REQUESTED** [review]

File: packages/frontend-api/src/Resources/config/services.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply packages/frontend-api/src/Resources/config/services.yaml:8]: thanks for remember.

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 15]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
     - also see #project-base-diff for more information about changes needed to be done in your project
 - lock version of npm `jquery-ui` to version `1.12.1` in order to fix category sorting in admin ([#2558](https://github.com/shopsys/shopsys/pull/2558))
     - see #project-base-diff for more information about changes needed to be done in your project
+- GQL resolvers and mutations refactoring ([#2563](https://github.com/shopsys/shopsys/pull/2563))
+    - all GQL Resolvers are renamed to Query:
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1194]: sure

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 16]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        - `Shopsys\FrontendApiBundle\Model\Resolver\Advert\AdverPositionsResolver` class are renamed:
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1195]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 17]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        ```diff
+        - Shopsys\FrontendApiBundle\Model\Resolver\Transport\TransportsResolver
+        + Shopsys\FrontendApiBundle\Model\Resolver\Transport\TransportsQuery
+        ```
+        - `Shopsys\FrontendApiBundle\Model\Resolver\Transport\TransportResolver` class are renamed:
+        ```diff
+        - Shopsys\FrontendApiBundle\Model\Resolver\Transport\TransportResolver
+        + Shopsys\FrontendApiBundle\Model\Resolver\Transport\TransportQuery
+        ```
+    - All functions in Query classes used as alias are renamed by convention with suffix `Query`
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1300]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 18]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        - in `Shopsys\FrontendApiBundle\Model\Resolver\Advert\AdverPositionsQuery` class is renamed function:
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1301]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 19]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        - public function resolveSearch(Argument $argument)
+        + public function categoriesSearchQuery(Argument $argument)
+        ```
+        - in `Shopsys\FrontendApiBundle\Model\Resolver\User\CurrentCustomerUserQuery` class is renamed function:
+        ```diff
+        - public function resolver(): CustomerUser
+        + public function currentCustomerUserQuery(): CustomerUser
+        ```
+        - in `Shopsys\FrontendApiBundle\Model\Resolver\Image\ImagesQuery` class are:
+          - replaced functions `resolveByCategory`, `resolveByPayment`, `resolveByTransport`, `resolveByBrand` for `imagesByEntityQuery`
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1364]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md [hunk 20]
```diff
@@ -1190,3 +1190,311 @@ There you can find links to upgrade notes for other versions too.
...
+        ```diff
+        - public function resolve(): array
+        + public function advertPositionsQuery(): array
+        ```
+        - in `Shopsys\FrontendApiBundle\Model\Resolver\Advert\AdversQuery` class in renamed function:
+        ```diff
+        - public function resolve(?string $positionName = null): array
+        + public function advertsQuery(?string $positionName = null): array
+        ```
+        - in `Shopsys\FrontendApiBundle\Model\Resolver\Article\ArticleQuery` class are renamed funcitons:
```
[the remark this answers is not in the document]
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1311]: ok

File: docs/frontend-api/introduction-to-frontend-api.md (hunk 1 above)
Comment (vitek-rostislav) [reply docs/frontend-api/introduction-to-frontend-api.md:167]: hm, if I had to choose, I would probably vote for Tom's suggestion, but I can live with both versions :wink:

File: upgrade/UPGRADE-v11.0.0-dev.md (hunk 10 above)
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1440]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md (hunk 11 above)
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1445]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md (hunk 12 above)
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1450]: oki

File: upgrade/UPGRADE-v11.0.0-dev.md (hunk 13 above)
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md:1457]: ok

File: upgrade/UPGRADE-v11.0.0-dev.md (hunk 14 above)
Comment (RostislavKreisinger) [reply upgrade/UPGRADE-v11.0.0-dev.md]: ok, i added all changes

**Review by TomasLudvik — APPROVED** [review]

**Review by pk16011990 — APPROVED** [review]

**Pull request merged**